### ISSUE TITLE
Bounding box filtering only respects clipping

### ISSUE THREAD

**Issue by jangop** [albumentations-team/albumentations#1322]
Bounding boxes that result from augmentations are filtered, such that bad/unsatisfactory bounding boxes can be removed. For this, `Compose` takes the argument `bbox_params` that takes an instance of `BboxParams`. The process is described in the documentation: https://albumentations.ai/docs/getting_started/bounding_boxes_augmentation/#min_area-and-min_visibility

Supposedly, using `min_visibility`, a bounding box is filtered out according to the ratio of the area before and after augmentation:

> `min_visibility` is a value between 0 and 1. If the ratio of the bounding box area after augmentation to _the area of the bounding box before augmentation_ becomes smaller than `min_visibility`, Albumentations will drop that box.

The actual filtering is performed via [`filter_bboxes`](https://github.com/albumentations-team/albumentations/blob/3904898e3bfe476137a75cad6ae3d1015e3d468e/albumentations/core/bbox_utils.py#L431):

```python
def filter_bboxes(
    bboxes: Sequence[BoxType], rows: int, cols: int, min_area: float = 0.0, min_visibility: float = 0.0
) -> List[BoxType]:
    ...
    resulting_boxes: List[BoxType] = []
    for bbox in bboxes:
        transformed_box_area = calculate_bbox_area(bbox, rows, cols)
        bbox, tail = cast(BoxType, tuple(np.clip(bbox[:4], 0, 1.0))), tuple(bbox[4:])
        clipped_box_area = calculate_bbox_area(bbox, rows, cols)
        if (
            clipped_box_area != 0  # to ensure transformed_box_area!=0 and to handle min_area=0 or min_visibility=0
            and clipped_box_area >= min_area
            and clipped_box_area / transformed_box_area >= min_visibility
        ):
            resulting_boxes.append(cast(BoxType, bbox + tail))
    return resulting_boxes
```

Areas are calculated before and after **clipping**. Width and height of the image do not change between the two calls to `calculate_bbox_area`. Therefore, **resizing is not taken into account**. This conflicts with the documentation and possibly the intent of the authors.

Fixing this is going to require either keeping the original bounding boxes and image size around, or precomputing and storing each bounding box's area. The alternative is simply a change to the documentation:I believe that it is not strictly necessary to filter based on visibility with regards to scaling, since there is still `min_area`.

However, bounding boxes with very small widths or heights can cause issues. A very tall bounding box, for example, might end up with a width considerably below even one pixel, but still boast an area of well over `200`. I would propose adding `min_width` and `min_height` to `filter_bboxes` and `BboxParams`. With trivial default values, this should be entirely backwards-compatible.

One last point to add: there is also [`filter_bboxes_by_visibility`](https://github.com/albumentations-team/albumentations/blob/3904898e3bfe476137a75cad6ae3d1015e3d468e/albumentations/core/bbox_utils.py#L236) which seems to have been written with similar intent to `filter_boxes`, but is entirely untested and with a different signature. Can anyone recall why `filter_bboxes_by_visibility` exists?

**Comment by Dipet** [comment]
You are right. I think we could add tracking shape before and after by add this logic [there](https://github.com/albumentations-team/albumentations/blob/master/albumentations/core/composition.py#L213) and [there](https://github.com/albumentations-team/albumentations/blob/master/albumentations/core/composition.py#L217).
In first case we need to save shapes before transform call. In second case - save shapes at the top of `__call__`

> One last point to add: there is also [filter_bboxes_by_visibility](https://github.com/albumentations-team/albumentations/blob/3904898e3bfe476137a75cad6ae3d1015e3d468e/albumentations/core/bbox_utils.py#L236) which seems to have been written with similar intent to filter_boxes, but is entirely untested and with a different signature. Can anyone recall why filter_bboxes_by_visibility exists?

It looks like legacy, because I do not see any usage of this function.

**Comment by jangop** [comment]
> You are right. I think we could add tracking shape before and after by add this logic [there](https://github.com/albumentations-team/albumentations/blob/master/albumentations/core/composition.py#L213) and [there](https://github.com/albumentations-team/albumentations/blob/master/albumentations/core/composition.py#L217). In first case we need to save shapes before transform call. In second case - save shapes at the top of `__call__`

Great suggestions. I do not know the code well enough to form a preference. 🤷 From my standpoint, either option seems fine.

> It looks like legacy, because I do not see any usage of this function.

👍 Does albumentations have a release schedule? Any established way to mark the function as deprecated and scheduled for removal?

**Comment by Dipet** [comment]
> Does albumentations have a release schedule?

No, we don't have it.

> Any established way to mark the function as deprecated and scheduled for removal?

This function is not a transform and made for internal use. So we could remove it with any next PR with code refactoring.
In common, we trying do not remove any transform. We just add a deprecation warning and hide it from the documentation.
